lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -559,6 +559,147 @@ namespace Js
+                catchFinally = library->CreatePromiseThenFinallyFunction(EntryThenFinallyFunction, RecyclableObject::FromVar(args[1]), constructor, true);
+            }
+            else
+            {
+                thenFinally = args[1];
+                catchFinally = args[1];
+            }
+        }
+        else
+        {
+            thenFinally = library->GetUndefined();
+            catchFinally = library->GetUndefined();
+        }
+
+        //7. Return ? Invoke(promise, "then", « thenFinally, catchFinally »).
+        Var funcVar = JavascriptOperators::GetProperty(promise, Js::PropertyIds::then, scriptContext);
+        if (!JavascriptConversion::IsCallable(funcVar))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedFunction, _u("Promise.prototype.finally"));
+        }
+        RecyclableObject* func = RecyclableObject::UnsafeFromVar(funcVar);
+
+        return CALL_FUNCTION(scriptContext->GetThreadContext(),
+            func, Js::CallInfo(CallFlags_Value, 3),
+            promise,
+            thenFinally,
+            catchFinally);
+    }
+
+    //ThenFinallyFunction as described in draft of ES2018  section 25.4.5.3.1
+    //AND CatchFinallyFunction as described in draft of ES2018  section 25.4.5.3.2
+    Var JavascriptPromise::EntryThenFinallyFunction(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        ScriptContext* scriptContext = function->GetScriptContext();
+        
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        JavascriptPromiseThenFinallyFunction* This = JavascriptPromiseThenFinallyFunction::FromVar(function);
+
+        //1. Let onFinally be F.[[OnFinally]]
+        //2. Assert: IsCallabale(onFinally)=true
+        Assert(JavascriptConversion::IsCallable(This->GetOnFinally()));
+
+        //3. Let result be ? Call(onFinally, undefined)
+        Var result = CALL_FUNCTION(scriptContext->GetThreadContext(), This->GetOnFinally(), CallInfo(CallFlags_Value, 1), library->GetUndefined());
+
+        //4. Let C be F.[[Constructor]]
+        //5.Assert IsConstructor(C)
+        Assert(JavascriptOperators::IsConstructor(This->GetConstructor()));
+
+        //6. Let promise be ? PromiseResolve(c, result)
+        Var promiseVar = CreateResolvedPromise(result, scriptContext, This->GetConstructor());
+
+        //7. Let valueThunk be equivalent to a function that returns value
+        //OR 7. Let thrower be equivalent to a function that throws reason
+        JavascriptPromiseThunkFinallyFunction* Thunk = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, args[1], This->GetShouldThrow());
+
+        //8. Return ? Invoke(promise, "then", <<valueThunk>>)
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(promiseVar);
```

Comment:
Are we sure `UnsafeFromVar` is ok here? Maybe regular `FromVar` is better unless `CreateResolvedPromise` has a guarantee.

Reply:
I think this is fine - CreateResolvedPromise uses a FromVar on the constructor to check it's safe then uses it to generate the promise. For this promise to be invalid therefore without hitting the FailFast in CreateResolvedPromise would require CreateResolvedPromise to be broken. - Not changed - will change if you still want me to though.

Reply:
The `FromVar` in `CreateResolvedPromise` is done to the `constructor` object, though. Here we are converting the promise object which was returned from `CreateResolvedPromise` which could be any `Var`, couldn't it? It's just whatever was returned from the `constructor` when we called it in `CreatePromiseCapabilityRecord`. Since we're calling as-constructor, the return should be an object, I guess. I do wonder if it would be possible to get it to be a non-object. Seems might be possible. I think we should be cautious and follow the `Invoke` rabbit-hole down to 7.3.2 GetV(V,P) ([https://tc39.github.io/ecma262/#sec-getv](url)) where we should do `ToObject` on `promise`. We do something very similar in `EntryAll` on a newly constructed promise Var.

Reply:
OK I've put in an AssertOrFailFast for it being an object, I think that should do the job? If not, not quite sure what you wanted.

Reply:
@rhuanjl I think he just wanted you to change the UnsafeFromVar to a FromVar.

Reply:
Sorry for the confusion - I said a lot of words.

When I saw this code at first I wondered if UnsafeFromVar would be actually safe because promiseVar is a Var and it isn't clear that it always must be a RecyclableObject. That's why I thought FromVar would be safer because it will check at runtime.

After taking a look further, I believe it may be possible for CreateResolvedPromise to return a non-object because the constructor may be overridden via Symbol.species. If that's the case, we should actually throw a TypeError while doing Invoke according to ECMA. Invoke calls GetV which performs ToObject on V. Does that make sense? 🤔

I believe the conversion should look like this:
```
RecyclableObject* promise = JavascriptOperators::ToObject(promiseVar, scriptContext);
```

Reply:
Yep that makes sense - well that aspect of the spec does not seem sensible at all but ah well. Will re-push after re-building and re-running tests.

Reply:
Many aspects of the spec do not seem sensible. Thanks for your efforts. 😄